#3543 Refactored UI Client and Server for Development

jmmadsen opened the pull request:
Fingers crossed this PR is correct this time...

@planetf1  I corrected the diff, so that only the appropriate files are shown changed. Only 163 files time around, as opposed to 2000+. Hoping we can discuss Maven considerations after this PR.

@davidradl I corrected the .gitignore according to your guidance. Unfortunately, I do not have time before the end of next week to pull in your latest changes. I am hoping this PR can be accepted so that the "structure" of the of the development server is at least set up. That way, others with more time can update the code with your latest changes (or I can get to it next week) and then begin using the easier development mode.

I referenced my issues with Chrome in the original issue #3543 . I created this one as a draft PR just to be safe.

jmmadsen commented:
@dkneisly If this PR is accepted, we could work together on correcting the discrepancies between the old client code and what I have updated.

planetf1 commented:
I noticed the file '.mentation/user-interfaces/presentation-server/cra-client/public/egeria_logo.png' ... I wonder if we could re-use one of the existing files already checked in to the build rather than duplicating?

planetf1 commented:
I do agree it makes sense to get the basics done & this PR cleared before we get too concerned with maven/gradle integration.

I'd be happy to give the build a go next week.

jmmadsen commented:
For `egeria_logo.png`, I chose that because I did not like how any of the others looked (I did try them as well). The image goes in the top lefthand corner of the browser tab so it is tiny, and the teal image is easier to make out than the others. However, I do not feel strongly about this and can change it if you like.

As for the rest of the images, that question is probably better directed towards @davidradl. They were chosen before I joined the project.

planetf1 commented on .gitignore:
```diff
@@ -115,6 +115,20 @@ open-metadata-implementation/user-interfaces/presentation-server/nodejs/yarn.loc
 open-metadata-implementation/user-interfaces/presentation-server/nodejs/node/**
 open-metadata-implementation/user-interfaces/presentation-server/nodejs/.env
 
+open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules
```
Quite a number of these files are already excluded by pattern (build, .vscode, *.log). Others may be included by minor modification to the existing patterns (package-lock?)

Not critical for this pr, so could be in a later one, though I can take a look to suggest more specifics if I get the overall (dev - not maven!)  working for the PR in my environment

jmmadsen replied:
This makes sense, but I like your suggestion for leaving it to a later PR. Perhaps we can discuss in our meeting next week

planetf1 replied:
#3575 opened to track

jmmadsen replied:
Addressed in #3577

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/README.md:
```diff
@@ -38,14 +38,6 @@ the types of [Egeria OMAG Servers](../../admin-services/docs/concepts/omag-serve
  called .env in the same folder and amend the values as required.
  * Once the .env file is in place, build and start the server using `npm start`. Use other npm parameters as specified in [package.json](nodejs/package.json).  
-
-## FOR NEW DEVELOPMENT (as of 8/2/2020)
```
May be an idea to use ISO format dates YYYY-MM-DD as we have a global audience and this is US date format

jmmadsen replied:
changed date format

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/README.md:
```diff
@@ -38,14 +38,6 @@ the types of [Egeria OMAG Servers](../../admin-services/docs/concepts/omag-serve
- * Because of this, work is currently underway to remake the client with Create React App. This will be done iteratively, and updates posted here.
- * Old client code will be maintained until the cutover is complete, so developers can run the traditional way if desired.
- * The new client is maintained in the cra-client directory.
- * Set up your environment by running `yarn`.
```
presumably from this directory - but also would require running yarn? might be worth adding a link even for this stage - though when we LATER do maven integration it shouldn't be needed (there we need to minimize any external dependencies)

jmmadsen replied:
We are actually not using Yarn anymore. I had it in the first PR I had merged, but after discussing further with @davidradl we decided to stick with NPM instead and have removed all Yarn references.

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/README.md:
```diff
@@ -54,6 +46,16 @@ the types of [Egeria OMAG Servers](../../admin-services/docs/concepts/omag-serve
   then the browser url to use to login is `https://localhost:8091/aaa`. All browser requests for this tenant will be issued using urls starting
   `https://localhost:8091/aaa`. 
     
+## FOR NEW DEVELOPMENT (as of 8/25/2020)
```
ISO date format

jmmadsen replied:
fixed

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/README.md:
```diff
@@ -54,6 +46,16 @@ the types of [Egeria OMAG Servers](../../admin-services/docs/concepts/omag-serve
+ * Because of this, work is currently underway to remake the client with Create React App. This will be done iteratively, and updates posted here.
+ * Old client code will be maintained until the cutover is complete, so developers can run the traditional way if desired.
+ * The new client is maintained in the cra-client directory.
+ * The new server is maintained in the cra-server directory.
```
should be careful of terminology as we use 'server' to mean an egeria server, yet here I think you mean a 'process listening on a port' ie network server. We use the word platform. Not sure though that is best here. but client/server can be confusing given how much we have this all over egeria... 
(Note -- it's ok we refine the docs in a later PR - just general feedback)

jmmadsen replied:
I tried to clarify a little further in the Readme, but agreed we can refine the verbiage as we go on

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/README.md:
```diff
@@ -54,6 +46,16 @@ the types of [Egeria OMAG Servers](../../admin-services/docs/concepts/omag-serve
+ * Set up your environment by running `npm install` in both cra-client and cra-server.
+ * Start the new UI by navigating into the cra-server directory, then running `npm start`.
+ * Enter `https://localhost:3000/` to access the new UI in development mode.
+ * Due to https, was only able to get it running properly using Firefox browser (no Chrome).
```
What issues did you get? I will try later... it may be due to the way current versions of chrome do further validation. We hit some issues even for the base platform. Whilst the config work isn't complete, it does work (in chrome) and notes are at #1403. Specifically it could relate to the certificate you've used. 

One option -- if we don't fix during PR, open up a new issue & fix it there

jmmadsen replied:
We have a separate conversation regarding this in #3543, so will move topic to there.

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-client/.env.sample:
```diff
@@ -0,0 +1,8 @@
+
+# This file is a sample .env file that can be used by developers.
+# Copy this file to a file called .env and modify the values as required
+HTTPS=true
```
be good to add comments on what the files are for (I realise, but for benefit of others)

jmmadsen replied:
add clarifying comments

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-client/package.json:
```diff
@@ -30,7 +30,8 @@
   "author": "Jacob Madsen",
-  "license": "ISC",
+  "license": "Apache-2.0",
+  "proxy": "https://localhost:8091",
```
I've not tried the code - presume this will need changing when deployed?

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-client/public/index.html:
```diff
@@ -5,7 +5,7 @@
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico">
+    <link rel="shortcut icon" href="%PUBLIC_URL%/egeria_logo.png">
```
Can we reuse one of the existing egeria logos and keep the files in the same place, so that our branding is kept together?

If requried the extra file can be brought into the build tree by a maven step once we integrate the build (getting resources)

jmmadsen replied:
@planetf1 How important would you say it is to keep this file in the same directory as all the others? It is a bit of pain to work with references in the `index.html` in React (but doable).

I do think we should use the new logo. I think the color stands out the best on the browser tab compared to others, which are hard to make out.

planetf1 replied:
I don't like the idea of duplication. the project as a whole should be coherent. The standard way of doing this would be - if needed - to copy in the required resources using maven (or gradle), probably better than relative links which can be problematic.

As such I suggest we defer to a seperate issue to resolve - leave as is for now.

planetf1 replied:
see #3576

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-client/src/Frame.js:
```diff
@@ -1,15 +1,17 @@
 import HeaderContainer from "carbon-components-react/lib/components/UIShell/HeaderContainer";
 import { Content } from "carbon-components-react/lib/components/UIShell";
 import { Link, Route, Switch } from "react-router-dom";
 import Egeriawhite110 from "./images/Egeria_logo_white_110";
```
as above - can we reuse these icons from egeria core

jmmadsen replied:
Do you mind referencing where these are stored in the egeria core? I am not super familiar with the rest of the project yet. I see `assets/img`, but that directory does not seem to have these assets.

planetf1 replied:
good question ... my thought was to collect them together - so add if needed, but needs further discussion so let's consider in #3576

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-client/src/contexts/IdentificationContext.js:
```diff
@@ -16,31 +17,28 @@ const IdentificationContextProvider = props => {
+      serviceName +
+      "/users/" +
+      userId;
+      console.log("rest url is ", url);
```
is this intended for developer debugging or the end user debugging (both)? This looks ok - but need to be careful not to introduce any personally identifiable info (that isnt already obvious from the browser env)

jmmadsen replied:
@davidradl can you provide any input? I did not develop this piece.

davidradl replied:
@planetf1 - this is part of the middleware that dos the multi-tenant support. I suggest we do not review the changes to the source code here unless there is a side effect .

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/.env_sample:
```diff
@@ -0,0 +1,10 @@
+# Each line configures a UI server (tenant) with the appropriate downstream server
+#    
+# In this example the first line configures a ui server called aaa with a remote server name of "cocoView1" and URL of "https://localhost:9443"
+EGERIA_PRESENTATIONSERVER_SERVER_aaa={"remoteServerName":"cocoView1","remoteURL":"https://localhost:9443"}
```
When we build an image of this in a container, it should be easy to get this env var -- presumably that is all you are doing with .env. -- initializing the environment?

jmmadsen replied:
Yes, I believe this is correct

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/index.js:
```diff
@@ -0,0 +1,70 @@
+
+
+// ssl self signed certificate and key
+const cert = fs.readFileSync(path.join(__dirname, '../') + "/../presentation-server/ssl/keys/server.cert");
```
this is probably something we'll need to return to later (so we have consolidated certs/clear docs)- but should go with this for now

jmmadsen replied:
agreed

planetf1 replied:
let's handle this in #1403

jmmadsen replied:
#1403

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/index.js:
```diff
@@ -0,0 +1,70 @@
+// create the https server
+https.createServer(options, app).listen(PORT, () => {
+  console.log(`Server listening on port: ${PORT}`);
+  if (env === 'development') console.log(`React UI listening on port: 3000`);
```
the port 3000 is hardcoded in log msg - should be ${PORT} ?

jmmadsen replied:
Line 68 has `${PORT} to allow for a dynamic change on the server. `3000` refers to the port of the UI, which is started in parallel and runs on a different port. Agreed it is not ideal to hardcode this, but not sure of a simple method to dynamically determine the port of another app.

planetf1 replied:
maybe just reword to 'react UI listening on port (default 3000)' just to be clear, otherwise it's potentially misleading.
The ports will need to be configurable

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/package.json:
```diff
@@ -0,0 +1,28 @@
+    "build": "cd ../cra-client && npm run build",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "Jacob Madsen",
```
We should probably list it as 'Egeria contributors' or similar, we we do with the copyright

jmmadsen replied:
Agreed, updated both package.json files

planetf1 replied:
thanks

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/router/routes.js:
```diff
@@ -0,0 +1,275 @@
+    instance
+      .get()
+      .then(function (response) {
+        console.log("response");
```
we may need to guard this or remove it if any potential for personally identifiable information

planetf1 replied:
is anything in the response sensitive in any way? Who sees this log? is it just the user accessing the ui? or the server admin? The former is ok, the latter requires care.

planetf1 replied:
@davidradl ?

jmmadsen replied:
Commented out lines in question

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/ssl/keys/server.cert:
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
+MIIDnjCCAoYCCQDR7DKQq8tpSzANBgkqhkiG9w0BAQsFADCBkDESMBAGA1UECAwJ
```
see comment above about certs - if we ship one may be useful to doc how. but can be later PR..

planetf1 replied:
as above let's handle in #1403

jmmadsen replied:
Agreed. Issue addressed in #1403

planetf1 reviewed (changes requested):
Tried to add some useful comments - it doesn't mean we need to fix all in this PR though

jmmadsen commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/db/users.js:
```diff
@@ -0,0 +1,38 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright Contributors to the ODPi Egeria project. */
+var records = [
```
added all the users referenced in that issue

jmmadsen commented:
Addressed some of Nigel's feedback above, and started dialogue on others. I believe the ones I did not immediately change can be addressed in future PRs as well.

I have changed this PR status from draft to ready for review.

planetf1 commented:
Thanks - I've updated most of the comments. I think it was just the question about taking care with logging / personal info, but I don't know the UI enough to know if that's an issue. We have had an issue reported previously in our other UI of this nature.

davidradl commented:
@jmmadsen I am seeing errors in the cra-client.

“davidradley@admins-MBP presentation-server % cd cra-client
davidradley@admins-MBP cra-client % npm install
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN deprecated @hapi/joi@15.1.1: joi is leaving the @hapi organization and moving back to 'joi' (https://github.com/sideway/joi/issues/2411)
npm WARN deprecated @hapi/bourne@1.3.2: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/topo@3.1.6: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/address@2.1.4: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/hoek@8.5.1: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated left-pad@1.3.0: use String.prototype.padStart()
npm WARN deprecated core-js@2.6.11: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.

> fsevents@1.2.13 install /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/jest-haste-map/node_modules/fsevents
> node install.js

No receipt for 'com.apple.pkg.CLTools_Executables' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLILeo' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLI' found at '/'.

gyp: No Xcode or CLT version detected!
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:311:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/Cellar/node/13.7.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/jest-haste-map/node_modules/fsevents
gyp ERR! node -v v12.16.1
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok 

> fsevents@1.2.13 install /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/watchpack-chokidar2/node_modules/fsevents
> node install.js

No receipt for 'com.apple.pkg.CLTools_Executables' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLILeo' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLI' found at '/'.

gyp: No Xcode or CLT version detected!
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:311:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/Cellar/node/13.7.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/watchpack-chokidar2/node_modules/fsevents
gyp ERR! node -v v12.16.1
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok 

> fsevents@1.2.13 install /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/webpack-dev-server/node_modules/fsevents
> node install.js

No receipt for 'com.apple.pkg.CLTools_Executables' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLILeo' found at '/'.

No receipt for 'com.apple.pkg.DeveloperToolsCLI' found at '/'.

gyp: No Xcode or CLT version detected!
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:311:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/Cellar/node/13.7.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/webpack-dev-server/node_modules/fsevents
gyp ERR! node -v v12.16.1
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok 

> node-sass@4.14.1 install /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/node-sass
> node scripts/install.js

Cached binary found at /Users/davidradley/.npm/node-sass/4.14.1/darwin-x64-72_binding.node

> core-js@2.6.11 postinstall /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/babel-runtime/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> https://opencollective.com/core-js 
> https://www.patreon.com/zloirock 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> core-js@3.6.5 postinstall /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> core-js-pure@3.6.5 postinstall /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/core-js-pure
> node -e "try{require('./postinstall')}catch(e){}"


> node-sass@4.14.1 postinstall /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/node-sass
> node scripts/build.js

Binary found at /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/node-sass/vendor/darwin-x64-72/binding.node
Testing binary
Binary is fine
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN notsup Unsupported engine for watchpack-chokidar2@2.0.0: wanted: {"node":"<8.10.0"} (current: {"node":"12.16.1","npm":"6.13.4"})
npm WARN notsup Not compatible with your version of node/npm: watchpack-chokidar2@2.0.0
npm WARN tsutils@3.17.1 requires a peer of typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta but none is installed. You must install peer dependencies yourself.

added 1838 packages from 835 contributors and audited 1842 packages in 47.876s

67 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
“

davidradl commented:
@jmmadsen I see 2 issues I think we should address before merging:
1. I get errors in the cra client build output and lots of warning of deprecations. I think we need to clean these up before merge as I cannot tell if there are any real errors there. 
2. I cannot login on Chrome or Safari, the issue appears to be related to cors.  See https://daveceddia.com/access-control-allow-origin-cors-errors-in-react-express/ for ways round this. I think this is a pretty fundamental we should address before the merge - so I can test whether it works. I am asking around, but It seems that in dev mode we could accept requests from all origins - this seems a reasonable approach. I wonder whether you have ever hit this in your previous projects?

jmmadsen commented:
@davidradl That makes sense to me. We can discuss further in our call tomorrow. Quick comments before that:

1. I assume these are linting errors, and did not actually prevent you from running the UI? Were you able to test basic connectivity to OMAG servers still worked in Firefox?
2. I have seen this before and can work on it next. Not sure when I will get to it this week though.
3. Easy fix.

planetf1 commented:
@davidradl I added a ref in #3543 to #1403. I did fix in the original PR, but also the new one in #3573 where I recorded the commands I used to generate the keys at https://github.com/odpi/egeria/pull/3238/files#diff-20ccd470941ca55f762c8aad9cba3d85 . Specifically I found it important to include the subjectAltName , as well as supply the correct basicConstraints, to get the very latest versions of chrome working. Further I aimed to include TLSv3 support -- and saw an error with firefox complaining about TLSv2. That may have been code related.

Apologies I was bogged down fixing blocking issues on the last release. My intent was to try swapping the certs, but I've not done so yet. Maybe this isn't the issue -- but I don't know what commands were used to generate the certs used in the UI code.

planetf1 commented:
On the cert/key itself... I wonder if in future we should use pkcs#12 format. For the java based code we had too options - the java format, or pkcs12. I chose the latter as more standard cross-language, so it would make things a lot easier if we used it here.

As before no reason to hold up the merge for this, but it would be better to consolidate if we can.

davidradl commented:
@jmmadsen 
1)I am not sure they are linting errors; they looked like dependancy issues. I suggest switching off the linter for now and getting it clean.  
2)Ok sounds good. Also could you include instructions on how I can test as a production build; I think we thought this was "for free" asa consequence of using Create React App.  
3) Sounds good.

jmmadsen commented:
@davidradl Finally had time to pick this up again after vacation. I went through and update all vulnerabilities in cra-react and cra-server directories, so NPM libraries should be good. I also added in the package.json files. If there is any logging still is isn't relevant to vulnerabilities. 

I was also able to get the development version running in Chrome. I added the .cert file to my keychain, and it allowed me to bypass the security warning. I believe this warning will persist locally as long as we are using a self-signed certificate.

Lastly, I do not have the build working yet. I started working on it but thought it would be best to get these changes in first, as this PR has been sitting stale for a while. We can discuss in our next meeting if you like.

planetf1 commented:
FYI I fixed the chrome cert issue in a recent PR https://github.com/odpi/egeria/pull/3684 and specifically updated the certs in 5d9552ae6d014e6851c0e1c33813e3d7dbadc199

With these in place the self signed warning can be bypassed. Without that's not possible as there were attributes missing in the certificate (there's a long discussion in #1403

planetf1 commented:
Also one request. As we don't have automated testing in place, but we do now include the UI in our notebookj environment/demos, can you ensure that still works from time to time.

Enhancements are another thing, but at a minimum we want to not break

Most likely it will be changes to layout/build that will break that.

The build does a 'mvn clean install' but the end result of a complete build on the build  machine is a new docker image. 

This can be run with
```
docker run odpi/egeria-presentation-server:latest%
```
or more precisely
```
docker run -p 8091:8091 -e EGERIA_PRESENTATIONSERVER_SERVER_coco="{ \"remoteServerName\":\"cocoView1\",\"remoteURL\":\"localhost:9444\"}"  odpi/egeria-presentation-server:latest
```
though this will run the most recent master image from our merge build - this takes up to 40 mins or so after a merge is complete

No PR support yet sadly - would be nice (#1734)

planetf1 commented:
@grahamwallis was that 'just' in my comment? I did find this bug and fix it in the code (we discussed it)

planetf1 commented:
Is this PR 'ready'? 
If so would it help if I took the PR (as it's a significant refactoring), and checked if things still work as they did in 2.3. I can try and writeup some better notes and share, so that we can figure out how we ensure we keep the deployment in sync moving forwards.

jmmadsen commented:
@planetf1 Met and discussed with @davidradl today to discuss the readiness. I believe it is ready, I think David is going to look it over to confirm.

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/ssl/keys/server.key:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
this is reintroducing a different key - we've just fixed this for 2.3 and are using the same key as the rest of the UI - for now. recommend this change is removed

planetf1 replied:
or at least corrected to use same file
it's also possible to copy the file by using the maven resources plugin in the pom - but that can be an optimization for later when we finess security

jmmadsen replied:
updated key to use one in master branch

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/ssl/keys/server.cert:
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
+MIIDnjCCAoYCCQDR7DKQq8tpSzANBgkqhkiG9w0BAQsFADCBkDESMBAGA1UECAwJ
+SGFtcHNoaXJlMRAwDgYDVQQHDAdIdXJzbGV5MRkwFwYDVQQKDBBMaW51eCBGb3Vu
```
this is reintroducing a different key - we've just fixed this for 2.3 and are using the same key as the rest of the UI - for now. recommend this change is removed

jmmadsen replied:
updated key to use one in master branch

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/package.json:
```diff
@@ -0,0 +1,33 @@
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "contributors": [
+    {
```
This should have the egeria contributor statement in not individuals

planetf1 replied:
search for 'webpack-build' in the pom.xml. This shows that currently we run webpack with the arguments `--mode development --devtool inline-source-map` in the nodejs directory. If this no longer works, those args/executions will need updating accordingly.

the file `open-metadata-distribution/open-metadata-assemblies/src/main/assemblies/egeria-presentation-server.xml`

then controls what gets put into an archive, which is then installed into a docker container which is defined in `open-metadata-resources/open-metadata-deployment/docker/presentation-server` . It is placed in /home/node/presentation-server & we start by running `node src/server/index.js`

planetf1 replied:
this is for a 'production' build. Obviously anything additional to aid in development turnaround is fine, and at least for a UI developer, I don't see the need to support in the POM (though possible if desired)

jmmadsen replied:
Updated contributor statement with standard Apache line. i am unsure about the second comment - perhaps we can discuss more after this merge? the production ready version of CRA is still not finalized yet, so i think this comment would be in line with those changes i plan on doing next

planetf1 commented on open-metadata-implementation/user-interfaces/presentation-server/cra-server/package.json:
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "cra-server",
```
is this the final name?

jmmadsen replied:
updated names to `presentation-server` and `presentation-server-client`

davidradl commented:
@jmmadsen I cant apply as the fix now has conflicts.I suggest going with the readme in master as it has been extensively reviewed

grahamwallis commented:
@planetf1 "was that 'just' in my comment?" Yes it was.

jmmadsen commented:
I have gone with the master ReadMe to work around the merge conflicts, and updated the majority of Graham's requested changes. If everyone is now okay with the merge conflicts, I wonder if we can merge this in and work on the webpack/docker/build related items for the next PR - where I plan on finalizing the build. In the meantime, Dave Kneisly will be able to benefit greatly from these changes to use hot fixes for his work.

davidradl commented:
@jmmadsen I hope to test later today or tomorrow.

sarbull commented on open-metadata-implementation/user-interfaces/presentation-server/cra-client/package.json:
```diff
@@ -24,13 +24,18 @@
   "devDependencies": {},
   "scripts": {
-    "start": "react-scripts start",
+    "start": "HTTPS=true react-scripts start",
```
in the official documentation, here [0], it says that in order to deploy the generated build, production code, you should set up your favourite HTTP server and they offer an example using "serve" or a nodejs with express static folder approach.

having this command [1] as a production command for starting the ui is actually starting a development server with HTTPS flag, adding `--prod` will also minify and uglify the code(as an example), but still, it is wrongly approached.

`react-scripts` should be used for development only and for bundling the code into production, like preparing the ENV variables and assets.

it also tells here [2] that, and i quote,
> This will simply start the basic, ExpressJS server included with the react-server package. It's not a total solution yet! Don't worry, we'll get there later in the guide.
which means there should be at least another layer of nginx as they say in the tutorial later.


[0] - https://create-react-app.dev/docs/deployment
[1] - `HTTPS=true react-scripts start`
[2] - https://react-server.io/docs/guides/production

jmmadsen replied:
Just pushed an update to remove the HTTPS flag for now. The correct build approach is going to be implemented in the next PR

jmmadsen replied:
The purpose of this PR is to implement react-scripts to enable more efficient development for the other UI authors before the build is finalized

davidradl commented:
@sarbull This is new code introducing Create React client - if you would like to feedback on it. It is work in progress and not complete and only works on Firefox currently; as we need a proxy to get round the security CORS errors. We are hoping to merge this pr this week

jmmadsen commented:
@davidradl For the record, I am able to get the most recent PR working in Chrome as well.

davidradl commented:
@jmmadsen I am giving Cesar a chance to review this. It worked for me on Firefox, but not on Chrome - I think this is expected.
I got 2 issues:
 1) I got an error during the client build "gyp: No Xcode or CLT version detected!
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:311:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/Cellar/node/13.7.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/jest-haste-map/node_modules/fsevents
gyp ERR! node -v v12.16.1
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok "

2. I also notice that after I build :
"	modified:   open-metadata-implementation/user-interfaces/presentation-server/cra-client/package-lock.json
	modified:   open-metadata-implementation/user-interfaces/presentation-server/cra-server/package-lock.json"
Appear in git status. These files should either be git ignored or included in the fix. I think we agreed to put them in the fix.

sarbull commented:
> @jmmadsen I am giving Cesar a chance to review this. It worked for me on Firefox, but not on Chrome - I think this is expected.
> I got 2 issues:
> 
> 1. I got an error during the client build "gyp: No Xcode or CLT version detected!
>    gyp ERR! configure error
>    gyp ERR! stack Error: `gyp` failed with exit code: 1
>    gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
>    gyp ERR! stack     at ChildProcess.emit (events.js:311:20)
>    gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
>    gyp ERR! System Darwin 19.6.0
>    gyp ERR! command "/usr/local/Cellar/node/13.7.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
>    gyp ERR! cwd /Users/davidradley/egeria/open-metadata-implementation/user-interfaces/presentation-server/cra-client/node_modules/jest-haste-map/node_modules/fsevents
>    gyp ERR! node -v v12.16.1
>    gyp ERR! node-gyp -v v5.0.5
>    gyp ERR! not ok "
> 
> 1. I also notice that after I build :
>    "	modified:   open-metadata-implementation/user-interfaces/presentation-server/cra-client/package-lock.json
>    modified:   open-metadata-implementation/user-interfaces/presentation-server/cra-server/package-lock.json"
>    Appear in git status. These files should either be git ignored or included in the fix. I think we agreed to put them in the fix.

You need to install CLT (command line tools) which comes with X Code from Mac.

I would personally also install "homebrew" which is like apt-get for ubuntu.

sarbull commented:
Updated the initial comment, added this one as a notification. Cheers!

jmmadsen commented:
I believe in my latest commit the `package-lock.json` is being tracked not ignored for the reasons @sarbull provided above.

davidradl commented:
Hi @jmmadsen without the instructions I thought I needed to do npm install so that generated new lock packages. I retested and realised I did not need to do the npm install.

planetf1 commented:
For reference by @sarbull to the node-gyp error above see https://github.com/odpi/egeria/issues/3761

We have to be very careful with platform specifics in the build -- especially as this module is part of the larger egeria build. Most people won't be touching the node UI (another reason for possibly breaking it out).. node-saas looks a bit problematic in this regard and urge 3761 is looked at

sarbull commented:
> For reference by @sarbull to the node-gyp error above see #3761
> 
> We have to be very careful with platform specifics in the build -- especially as this module is part of the larger egeria build. Most people won't be touching the node UI (another reason for possibly breaking it out).. node-saas looks a bit problematic in this regard and urge 3761 is looked at

@planetf1 most of the `node-sass` issues comes from the fact that it has to download some binary specifically for the used NODE version as  listed here [0] from an external URL which might not be visible in the build context.

the issue @davidradl had in regards to the node-gyp should be fixed by installing what i mentioned above.

[0] - https://github.com/sass/node-sass-binaries
